flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceSplitEnumerator.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.base.source.hybrid;
+
+import org.apache.flink.api.connector.source.ReaderInfo;
+import org.apache.flink.api.connector.source.Source;
+import org.apache.flink.api.connector.source.SourceEvent;
+import org.apache.flink.api.connector.source.SourceSplit;
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.connector.source.SplitsAssignment;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Callable;
+import java.util.function.BiConsumer;
+
+/**
+ * Wraps the actual split enumerators and facilitates source switching. Enumerators are created
+ * lazily when source switch occurs to support runtime position conversion.
+ *
+ * <p>This enumerator delegates to the current underlying split enumerator and transitions to the
+ * next source once all readers have indicated via {@link SourceReaderFinishedEvent} that all input
+ * was consumed.
+ *
+ * <p>Switching between enumerators occurs by either creating the new enumerator with a fixed start
+ * position via {@link Source#createEnumerator(SplitEnumeratorContext)} or by using the state of the
+ * previous enumerator and the optional user supplied checkpoint state converter to create the next
+ * enumerator with a runtime determined start position via {@link
+ * Source#restoreEnumerator(SplitEnumeratorContext, Object)}.
+ *
+ * <p>During subtask recovery, splits that have been assigned since the last checkpoint will be
```

Comment:
there are a lot of complexities to take care of this scenario. I am wondering if we can do the source switch only at  `notifyCheckpointComplete`. That might help avoid those complexities and the HybridSourceSplitEnumerator doesn't need to track any assignment or pendingSplits.

It has its own drawbacks
- delay the switch up to a checkpoint interval
- it requires checkpoint enabled. it won't work for batch sources that don't support or enable checkpoint. is it a valid concern?

Reply:
It is complex indeed but I think I still prefer the HybridSource to be able to function independent of checkpointing to avoid the limitations that you already mentioned (checkpointing not enabled, large intervals etc.).

Reply:
delayed cutover (up to a checkpoint interval) is probably not a big deal, as job can be delayed for up to a checkpoint interval during failure. Regarding the requirement of enabling checkpoint, that is where I am not too sure. The most common scenario for hybrid source is transitioning from historic data source (like file) to live data source (like Kafka), enabling checkpoints seem likely. However, I agree that it is better to avoid such a restriction.

I want to discuss the other implication of the current behavior though. It means that cutover/switch is not one-way final deal. In some failure cases, we may go back to the previous source after cutover. I am wondering what is the semantic of hybrid source. is it one-time final/irreversible action? that has implications on ordering. if the hybrid source can go back to the previous source after cutover, it is hard to achieve ordering guarantee.

Reply:
It follows the semantics of recovery from checkpoint. On a task manager failure, readers are reset, potentially going back before a switch. And the corresponding splits are added back to the enumerator. Then the splits that were added back are processed in the same order by the same subtasks. There shouldn't be any change with respect to the ordering. The ITCase covers that.

Reply:
> delayed cutover (up to a checkpoint interval) is probably not a big deal, as job can be delayed for up to a checkpoint interval during failure recovery.

For large intervals that would matter, because there would be a corresponding gap where the source does not emit any data while it is waiting for the next checkpoint to complete.)